CMakeLists.txt

```diff
@@ -57,6 +58,136 @@ option(ENABLE_CCACHE "Speed up recompilation by caching previous compilations" O
 
 # Platform-specific options
 
+if(EMSCRIPTEN)
+	if(NOT CMAKE_BUILD_TYPE STREQUAL "MinSizeRel")
+		message(FATAL_ERROR
+			"Only MinSizeRel is supported for emscripten"
+			"Use -DCMAKE_BUILD_TYPE=MinSizeRel when generate build configuration"
+		)
+	endif()
+	if(NOT UNIX)
+		message(FATAL_ERROR "Only UNIX build is supported for emscripten")
+	endif()
+
+	include(ExternalProject)
+	include(ProcessorCount)
+	ProcessorCount(N)
+
+	set(ENABLE_HTML5_BUILD ON)
+	ExternalProject_Add(boost-lib
+			PREFIX "boost"
+			URL "https://github.com/boostorg/boost/releases/download/boost-1.86.0/boost-1.86.0-cmake.zip"
+			BUILD_IN_SOURCE true
+			PATCH_COMMAND find . -name *.sh -execdir chmod u+x {} +
+			CONFIGURE_COMMAND ./bootstrap.sh
+			BUILD_COMMAND ./b2 install
+				--with-date_time
+				--with-filesystem
+				--with-locale
+				--with-program_options
+				--with-system
+				--with-thread
+				--with-chrono
+				--prefix=${CMAKE_BINARY_DIR}/boost
+				toolset=emscripten
+				link=static
+				variant=release
+				"-j${N}"
+			INSTALL_COMMAND ""
+			INSTALL_DIR ${CMAKE_BINARY_DIR}/boost
+	)
+	ExternalProject_Add(tbb-lib
+			PREFIX "tbb"
+			URL "https://github.com/oneapi-src/oneTBB/archive/refs/tags/v2022.0.0.zip"
+			BUILD_IN_SOURCE true
+			PATCH_COMMAND echo
+				"target_compile_options(tbb PRIVATE \"-fwasm-exceptions\")\\n"
+				"target_compile_options(tbbmalloc PRIVATE \"-fwasm-exceptions\")\\n"
+				"target_link_options(tbb PRIVATE \"-fwasm-exceptions\")\\n"
+				"target_link_options(tbbmalloc PRIVATE \"-fwasm-exceptions\")\\n"
+				>> CMakeLists.txt
+			CONFIGURE_COMMAND "emcmake"
+				"${CMAKE_COMMAND}"
+				"-G${CMAKE_GENERATOR}"
+				"-DCMAKE_INSTALL_PREFIX:PATH=${CMAKE_BINARY_DIR}/tbb"
+				"-DBUILD_SHARED_LIBS=false"
+				"-DCMAKE_BUILD_TYPE=MinSizeRel"
+				"-DCMAKE_CXX_FLAGS=-fwasm-exceptions"
+				"-DLINK_FLAGS=-fwasm-exceptions"
+			BUILD_COMMAND "${CMAKE_MAKE_PROGRAM}" "tbb" "tbbmalloc"
+			INSTALL_COMMAND "${CMAKE_MAKE_PROGRAM}" "install" "tbb" "tbbmalloc"
+	)
+	ExternalProject_Add(vcmi-extras
+			PREFIX "vcmi-extras"
+			GIT_REPOSITORY "https://github.com/vcmi-mods/vcmi-extras"
+			GIT_TAG "vcmi-1.6"
+			BUILD_IN_SOURCE true
+			CONFIGURE_COMMAND ""
+			BUILD_COMMAND ""
+			INSTALL_COMMAND ""
+	)
```

Comment:
Why is this needed for web version?

Reply:
I am not sure about it; it seems that vcmi-extras is something very commonly used. For example, most network games require it. It is used later to create a data file with required resources, like config. However, it is also true that the data can be bundled later and does not have to be part of the regular build. 

It is simpler to have this data built with client JS; then, you only need to provide the data from HoMM3.

Reply:
And a lot of players use hota. Bundle it as well? We have a lot of players from Poland. Bundle Polish translation as well?

Reply:
Make sense, will remove it

Reply:
having such stuff in project's CML is a very bad idea. Provide a script that downloads the necessary prerequisites (that only the web versions needs) before building.

Reply:
Do you mean vcmi-extras or all instances of ExternalProject_Add? Why is using it bad? It's cross-platform and works well.

Reply:
I'm talking about ExternalProject_Add. No other VCMI platform needs it and web version is an experimental thing atm, hence don't see a need to pollute CML with such things that are easily done with a custom script that would run before cmake invocation.